ldm/invoke/patchmatch.py

```diff
@@ -0,0 +1,43 @@
+'''
+This module defines a singleton object, "patchmatch" that
+wraps the actual patchmatch object. It respects the global
+"try_patchmatch" attribute, so that patchmatch loading can
+be suppressed or deferred
+'''
+from ldm.invoke.globals import Globals
+import numpy as  np
+
+class Patchmatch (object):
+    '''
+    Thin class wrapper around the patchmatch function.
+    '''
+    
+    def __init__(self):
+        self.patch_match = None
+        self.tried_load:bool = False
+        super().__init__()
+
+    def _load_patch_match(self):
+        if self.tried_load:
+            return
+        if Globals.try_patchmatch:
+            from patchmatch import patch_match as pm
+            if pm.patchmatch_available:
+                print('>> Patchmatch initialized')
```

Comment:
We should probably start using the actual logging functions in Python soon =)

Reply:
I'm using `logging` in the `generate` refactor that I started yesterday. I'll do a general search and replace for all the print() functions scattered around.

Sorry it's like that, but that's what happens when someone unfamiliar with Python style learns from the CompVis code.

Reply:
> general search and replace

we do need to make a distinction between logging and user messaging. Adding `logging` to `generate` (and elsewhere) is a great direction, but if I may ask to hold off on a sweeping  search-and-replace for `print()`, that would be preferable :) . 

`rich` is a great library for handling user messaging in console applications (we're already using it for prettifying stack traces via `traceback`) , and I started using it in `configure_invokeai` to better present some of the messaging. Hoping to have that PR ready soon

Reply:
We should make sure we use a logging library that can filter logs by category, and that we put some effort into what's the appropriate log level for each message. There's also some command line output (e.g. the progress bar) that's not really as necessary when running the web ui, and wouldn't be desired in a hosted environment (or if we parallelize).

Reply:
https://www.structlog.org is quite good and fits all of the above criteria. I've used it before, albeit only having scratched the surface of its full capabilities

Reply:
Let's put this out on the Discord for a conversation. I'm sure several people have a logging module they prefer.